### ISSUE TITLE
Proposal: Add support for custom parsers/extractors

### ISSUE THREAD

**Issue by OzairP** [palantir/tslint#4379]
### Feature request

**Is your feature request that we implement a new rule?**
No

**Is your feature request related to a problem? Please describe.**
Yes, inability to parse non-typescript files. E.g. parsing single file Vue components.

**Describe the solution you'd like**
Introduce a parsing/extraction step before the TypeScript compiler parses any kind of TS code.
This parser/extractor will extract the relevant source code from selected files and return an array of `Files`.
This file object looks something like
```ts
{
    // Relevant file name
    fileName: 'HomeComponent.vue',
    // Start position of relevant TS
    startPos: 349,
    // Source code
    source: '... ts source here ...'
}
```

Now in run time we extract the meaningful files like normal. Before we [`#doLinting`](https://github.com/palantir/tslint/blob/master/src/runner.ts#L184), we instead check which parser/extractor to use from the config file, by default pick a no-op parser/extractor.

It will run the parser/extractor, then pass all files to `#doLinting`, before we return we run through the errors and amend the `startPosition` and `endPosition` by adding `startPos`. (This can be mitigated by initially having the AST have correct start and end positions but I haven't thought of a good way to do that yet).

**Describe alternatives you've considered**
Using [Wotan](https://github.com/fimbullinter/wotan), but I believe this proposal isn't too complex and should be done without needing complex solutions.

**Additional context**
#2099

**Comment by JoshuaKGoldberg** [comment]
Thanks for bringing this up @OzairP! Agreed that the ability to parse files would be great, but there are some questions that'll need proposals & consensus:
* How would the API look?
* How can we get around https://github.com/Microsoft/TypeScript/issues/28413, a bug in which TypeScript compiler APIs tend to fail on manually created `ts.sourceFile`s not initially a part of the program?
* What are the supported or unsupported use cases?
    * What other languages like Vue have this kind of use case?
    * Would it be reasonable for users to use a Flow-to-TypeScript converter, for example?
* Would this compete with [TypeScript's language service plugins](https://github.com/Microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin)?

Let's spend some time seriously discussing the above committing to one. 

I also question whether this should exist in TSLint core itself. Vue already requires some kind of Babel/Webpack/etc. stage _(right?)_ so if you already need to use a tool like that to parse the files, what does adding it into TSLint itself give?

Also relevant: #4273 (`no-unsafe-any` bombs on .vue single-file-templates), which would also be fixed by a good solution here.

**Comment by JoshuaKGoldberg** [comment]
Given the lack of discussion or concrete proposal here and the upcoming deprecation of TSLint in favor of ESLint (#4534), I'm going to go ahead and close this issue and related Vue-specific reports for housekeeping purposes. 

If this is still an issue for you in [typescript-eslint](https://typescript-eslint), I'd recommend filing an issue there. Best of luck!
